src/main/java/org/jenkinsci/plugins/pipeline/milestone/MilestoneStepExecution.java

```diff
@@ -281,80 +168,83 @@ private static void println(StepContext context, String message) {
         try {
             context.get(TaskListener.class).getLogger().println(message);
         } catch (Exception x) {
-            LOGGER.log(WARNING, "failed to print message to dead " + context, x);
+            LOGGER.log(Level.WARNING, x, () -> "failed to print message to dead " + context);
         }
     }
 
-    private static void cancel(StepContext context, Integer build) throws IOException, InterruptedException {
-        if (context.isReady()) {
-            println(context, "Canceled since build #" + build + " already got here");
-            Run<?, ?> r = context.get(Run.class);
-            String job = "";
-            if (r != null) { // it should be always non-null at this point, but let's do a defensive check
-                job = r.getParent().getFullName();
+    @Extension
+    public static final class Listener extends RunListener<Run<?,?>> {
+        @Override
+        public void onStarted(Run<?, ?> r, TaskListener listener) {
+            if (isPipelineRun(r)) {
+                MilestoneStorage milestoneStorage = getStorage();
+                milestoneStorage.store(r, null);
             }
-            throw new FlowInterruptedException(Result.NOT_BUILT, new CancelledCause(job + "#" + build));
-        } else {
-            LOGGER.log(WARNING, "cannot cancel dead #" + build);
         }
-    }
 
-    private static void cleanUp(Job<?,?> job, String jobName) {
-        Map<Integer, Milestone> milestonesInJob = getMilestonesByOrdinalByJob().get(jobName);
-        assert milestonesInJob != null;
-        Iterator<Entry<Integer, Milestone>> it = milestonesInJob.entrySet().iterator();
-        while (it.hasNext()) {
-            Map.Entry<Integer, Milestone> entry = it.next();
-            Set<Integer> inSight = entry.getValue().inSight;
-            Iterator<Integer> it2 = inSight.iterator();
-            while (it2.hasNext()) {
-                Integer number = it2.next();
-                if (job.getBuildByNumber(number) == null) {
-                    // Deleted at some point but did not properly clean up from exit(…).
-                    LOGGER.log(WARNING, "Cleaning up apparently deleted {0}#{1}", new Object[] {jobName, number});
-                    it2.remove();
+        @Override public void onCompleted(Run<?,?> r, @NonNull TaskListener listener) {
+            if (isPipelineRun(r)) {
+                MilestoneStorage milestoneStorage = getStorage();
+                var result = milestoneStorage.clear(r);
+                MilestoneStorage.LOGGER.finest(() -> "milestones after completion: " + result.milestones());
+                if (result.lastMilestoneBeforeCompletion() != null) {
+                    MilestoneStorage.LOGGER.finest(() -> "Build" + r + " last milestone before completion: " + result.lastMilestoneBeforeCompletion());
```

Comment:
Prefer use of our own class’ `Logger`. Also this should probably be `FINE` or `FINER`.